[BUG] Fix databricks-aws user profiling tool error with `--gpu_cluster` argument

cindyyuanjiang opened the pull request:
Fixes https://github.com/NVIDIA/spark-rapids-tools/issues/695

**Problem**
The command `aws ec2 describe-instance-types --region us-west-2 --instance-types g5.8xlarge` returns GPU device name `A10G`, which is unrecognized by the profiling tool.

**Changes**
This PR adds a new type `A10G` in `GpuDevice` class.

amahussein reviewed (changes requested):
I wonder is this the only place we need to add the missing "A10G"?

What about other locations that look-up the GPU:

- [user_tools/src/spark_rapids_pytools/cloud_api/sp_types.py](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/user_tools/src/spark_rapids_pytools/cloud_api/sp_types.py#L52)
```
    P100 = 'P100'
    P4 = 'P4'
    L4 = 'l4'
    A10 = 'a10'
```

- [data_validation/src/spark_rapids_validation_tool/utilities.py](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/data_validation/src/spark_rapids_validation_tool/utilities.py#L113)

```
def get_gpu_device_list():
    return ['T4', 'V100', 'K80', 'A100', 'P100']
```

- [core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala#L980)

```
  // T4 default memory is 16G
  // A100 set default to 40GB
  val DEF_WORKER_GPU_MEMORY_MB: Map[String, String] = Map("T4"-> "15109m", "A100" -> "40960m")
```

- [user_tools/src/spark_rapids_pytools/resources/databricks_azure-configs.json](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/user_tools/src/spark_rapids_pytools/resources/databricks_azure-configs.json#L325)

```
          "GpuInfo": {
            "GPUs": [
              {
                "Name": "A100",
                "Manufacturer": "NVIDIA",
                "Count": 8,
                "MemoryInfo": {
```


- [user_tools/src/spark_rapids_pytools/resources/premium-databricks-azure-catalog.json](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/user_tools/src/spark_rapids_pytools/resources/premium-databricks-azure-catalog.json#L1556)

```
      "ND96asr_A100_v4": {
        "RAMinMB": 921600,
        "vCPUs": 96,
        "Instance": "ND96asr_A100_v4",
        "DBUCount": 44.0,
```

cindyyuanjiang commented:
> I wonder is this the only place we need to add the missing "A10G"?

Thanks @amahussein! I have updated the other use sites of GPU devices.

- [core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala#L980)

```
  val DEF_WORKER_GPU_MEMORY_MB: Map[String, String] = Map("T4"-> "15109m", "A100" -> "40960m")
```

For the map above, I have updated all the GPU devices based on [user_tools/src/spark_rapids_pytools/cloud_api/sp_types.py](https://github.com/NVIDIA/spark-rapids-tools/blob/8ed191a9b540ef8ac71ea9179d483e1dbcc00acd/user_tools/src/spark_rapids_pytools/cloud_api/sp_types.py#L58):

```
    def get_gpu_mem(self) -> list:
        memory_hash = {
            self.T4: [16384],
            self.L4: [24576],
            self.A100: [40960, 81920],
            self.P4: [8192],
            self.K80: [12288],
            self.V100: [16384],
            self.P100: [16384],
            self.A10: [24576]
        }
        return memory_hash.get(self)
```

We do not need to update the following sites because these info are pulled from databricks azure websites directly.

- [user_tools/src/spark_rapids_pytools/resources/databricks_azure-configs.json](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/user_tools/src/spark_rapids_pytools/resources/databricks_azure-configs.json#L325)

```
          "GpuInfo": {
            "GPUs": [
              {
                "Name": "A100",
                "Manufacturer": "NVIDIA",
                "Count": 8,
                "MemoryInfo": {
```


- [user_tools/src/spark_rapids_pytools/resources/premium-databricks-azure-catalog.json](https://github.com/NVIDIA/spark-rapids-tools/blob/40b1b9e74cb9a5302e0d46ff136a74d8af2a09ce/user_tools/src/spark_rapids_pytools/resources/premium-databricks-azure-catalog.json#L1556)

```
      "ND96asr_A100_v4": {
        "RAMinMB": 921600,
        "vCPUs": 96,
        "Instance": "ND96asr_A100_v4",
        "DBUCount": 44.0,
```

amahussein commented:
> We do not need to update the following sites because these info are pulled from databricks azure websites directly.

Thanks @cindyyuanjiang !
Are you sure there are no changes to be done in `databricks_aws-configs.json`? There is a list of machines there which does not include `g5` family.

amahussein commented:
And Probably `emr-configs.json` as well

cindyyuanjiang commented:
> Thanks @cindyyuanjiang ! Are you sure there are no changes to be done in `databricks_aws-configs.json`? There is a list of machines there which does not include `g5` family.

I double checked the use of `supportedGpuInstances` in `databricks_aws-configs.json` and `emr-configs.json`. This information is mainly used to find mapping from CPU instance types to GPU instance types. I think it would be a separate issue if we want to augment this list.

amahussein commented:
> > Thanks @cindyyuanjiang ! Are you sure there are no changes to be done in `databricks_aws-configs.json`? There is a list of machines there which does not include `g5` family.
> 
> I double checked the use of `supportedGpuInstances` in `databricks_aws-configs.json` and `emr-configs.json`. This information is mainly used to find mapping from CPU instance types to GPU instance types. I think it would be a separate issue if we want to augment this list.

The problem with the PR is that it only fixes a symptom.
The bug fix is that the device portfolio is out-dated and we should update it to prevent similar crashes with different platforms.

amahussein commented:
@cindyyuanjiang  
I guess that the issue #637  filed by @kuhushukla and her existing PR #638 is related to this very bug.
It looks like the Qualification crashed because it could not find the correct map on Databricks-Azure. That's why, I believe we should fix the root problem.

cindyyuanjiang commented:
> @cindyyuanjiang I guess that the issue #637 filed by @kuhushukla and her existing PR #638 is related to this very bug. It looks like the Qualification crashed because it could not find the correct map on Databricks-Azure. That's why, I believe we should fix the root problem.

Thanks @amahussein! Yes, I understand. This PR's issue is due to a missing GPU device type in class `GpuDevice`, instead of missing GPU instance types. If the Qualification tool fails because it cannot find mapping to GPU instance types, we should file a separate issue and fix the root cause. I am ok to fix everything in this PR too.

amahussein commented:
> > @cindyyuanjiang I guess that the issue #637 filed by @kuhushukla and her existing PR #638 is related to this very bug. It looks like the Qualification crashed because it could not find the correct map on Databricks-Azure. That's why, I believe we should fix the root problem.
> 
> Thanks @amahussein! Yes, I understand. This PR's issue is due to a missing GPU device type in class `GpuDevice`, instead of missing GPU instance types. If the Qualification tool fails because it cannot find mapping to GPU instance types, we should file a separate issue and fix the root cause. I am ok to fix everything in this PR too.

The primary issue is the absence of GPU devices, and we also lack maps for additional clusters.
Our approach to fix the bug involves enumerating all potential crash sites.  
The decision to split them into separate pull requests will be made based on the evaluation of workload and priorities.

To ensure this pull request (PR) is self-contained, please first file an issue as a follow-up, including the problem details and the suggested fix.

cindyyuanjiang commented:
Filed follow up issue for missing instance types in pricing information - https://github.com/NVIDIA/spark-rapids-tools/issues/729

amahussein commented on data_validation/src/spark_rapids_validation_tool/utilities.py:
Update copyrights  2023-2024

cindyyuanjiang replied:
Fixed, thanks

parthosa replied:
Thanks @cindyyuanjiang.  This file would also require copyright updates.

amahussein replied:
+1 @parthosa

cindyyuanjiang replied:
Now fixed, thanks

amahussein commented on user_tools/src/spark_rapids_pytools/cloud_api/sp_types.py:
Update copyrights 2023-2024

cindyyuanjiang replied:
Fixed, thanks

amahussein reviewed (changes requested):
Thanks @cindyyuanjiang 

- Copyrights updates should be added to all changed files
- Please make sure that the code is tested after a merge. For scala, at least the code should run `mvn clean package` successfully. For user-tools, at least one execution/runtime after the merge to make sure that the merge did not break the fixes.

cindyyuanjiang commented:
> * Copyrights updates should be added to all changed files
> * Please make sure that the code is tested after a merge. For scala, at least the code should run `mvn clean package` successfully. For user-tools, at least one execution/runtime after the merge to make sure that the merge did not break the fixes.

Thank you @amahussein 
I have updated the copyright years and tested with both `tox` and some command line runs.

parthosa commented:
Does the core tools also require `A10G` type in GpuTypes and Platform class?

amahussein commented on user_tools/src/spark_rapids_pytools/common/utilities.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2023, NVIDIA CORPORATION.
+# Copyright (c) 2023-2024, NVIDIA CORPORATION.
```
Looks like you updated this file's copyright instead of `data_validation/src/spark_rapids_validation_tool/utilities.py`

cindyyuanjiang replied:
sorry about this, now should be fixed.

amahussein reviewed (changes requested):
still incorrect copyrights

cindyyuanjiang commented:
> Does the core tools also require `A10G` type in GpuTypes and Platform class?

We have updated the core tools in a previous [PR](https://github.com/NVIDIA/spark-rapids-tools/commit/c0b4ddf0f230d8be6c26d9cfc725b2707e0eed96).

parthosa reviewed (approved):
Thanks @cindyyuanjiang
